dis_snek/gateway.py

```diff
@@ -39,177 +42,118 @@ async def rate_limit(self) -> None:
 
         self._trace = []
+        self.heartbeat_interval = None
+        self.sequence = None
+        self.session_id = None
 
-        self.closed = asyncio.Event()
-        self.resume = False
-        self.shutdown = False
+        self.latency = collections.deque(maxlen=10)
 
-        self.ready: asyncio.Event = asyncio.Event()
+        # This lock needs to be held to send something over the gateway, but is also held when
+        # reconnecting. That way there's no race conditions between sending and reconnecting.
+        self._race_lock = asyncio.Lock()
+        # Then this event is used so that receive() can wait for the reconnecting to complete.
+        self._closed = asyncio.Event()
+
+        self._keep_alive = None
+        self._kill_bee_gees = asyncio.Event()
+        self._last_heartbeat = 0
+        self._acknowledged = asyncio.Event()
+        self._acknowledged.set()  # Initialize it as set
+
+        # Santity check, it is extremely important that an instance isn't reused.
+        self._entered = False
 
     @property
     def loop(self):
-        return self.client.loop
-
-    @classmethod
-    async def connect(
-        cls,
-        state: "ConnectionState",
-        session_id: Optional[int] = None,
-        sequence: Optional[int] = None,
-        presence: Optional[dict] = None,
-    ):
-        """
-        Connect to the discord gateway
-        Args:
-            state: The connection state
-            session_id: The session id to use, if resuming
-            sequence: The sequence to use, if resuming
-            presence: The presence to login with
+        return self.state.client.loop
 
-        """
-        cls.client = state.client
-        cls.state = state
-        cls._gateway = await state.client.http.get_gateway()
-        cls.presence = presence
-        cls.ws = await state.client.http.websocket_connect(cls._gateway)
-
-        cls.shard_id = state.shard_id
-        cls.total_shards = state.client.total_shards
-        state.client.dispatch(events.Connect())
-        if session_id and sequence:
-            # resume
-            return cls(session_id, sequence)
-        return cls()
+    async def __aenter__(self: SELF) -> SELF:
+        if self._entered:
+            raise RuntimeError("An instance of 'WebsocketClient' cannot be re-used!")
 
-    @property
-    def latency(self) -> float:
-        """Get the latency of the connection."""
-        if self._keep_alive.latency:
-            return self._keep_alive.latency[-1]
-        else:
-            return float("inf")
+        self._entered = True
+
+        self.ws = await self.state.client.http.websocket_connect(self.state.gateway_url)
+        self._closed.set()
+
+        hello = await self.receive()
+        self.heartbeat_interval = hello["d"]["heartbeat_interval"] / 1000
+
+        await asyncio.sleep(self.heartbeat_interval * random.uniform(0, 0.5))
+        self._keep_alive = asyncio.create_task(self._start_bee_gees())
```

Comment:
This delays both heartbeating and indentify. Per api spec, only the heartbeat needs delaying by `heartbeat_interval * jitter`.

Reply:
Yes I originally read it like this but whilst implementing it here I realized that surely Discord meant both the heartbeating and IDENTIFY. Heartbeating itself can't be that costly for Discord - IDENTIFY is on the other hand.

[The documentation here](https://discord.com/developers/docs/topics/gateway#identifying) implies that after starting to heartbeat (which happens after that original delay) you should IDENTIFY - hence the wait delays both the heartbeat and IDENTIFY.

Reply:
I don't believe your interpretation is correct. You send your receive hello, then heartbeat_interval * jitter later you send heartbeats. I don't believe identify is included in that delay.